### ISSUE TITLE
runtime error  in sf16_to_sf32() in astcenc_mathlib_softfloat.cpp

### ISSUE THREAD

**Issue by fb-aki** [ARM-software/astc-encoder#286]
I am seeing the following runtime error:

```
runtime error: left shift of negative value -2147483648
```
at `astcenc_mathlib_softfloat.cpp:172` (this is in version 3.2)

It appears that the `tbl` defined above it should have unsigned integer values instead of signed integer values.  I've locally made the change:

```
	#define WITH_MB(a) uint32_t((a) | (1 << 31))
	static const uint32_t tbl[64] =
	{
		WITH_MB(0x00000), UINT32_C(0x1C000), UINT32_C(0x1C000), UINT32_C(0x1C000), UINT32_C(0x1C000), UINT32_C(0x1C000), UINT32_C(0x1C000), UINT32_C(0x1C000),
		UINT32_C(0x1C000), UINT32_C(0x1C000), UINT32_C(0x1C000), UINT32_C(0x1C000), UINT32_C(0x1C000), UINT32_C(0x1C000), UINT32_C(0x1C000), UINT32_C(0x1C000),
		UINT32_C(0x1C000), UINT32_C(0x1C000), UINT32_C(0x1C000), UINT32_C(0x1C000), UINT32_C(0x1C000), UINT32_C(0x1C000), UINT32_C(0x1C000), UINT32_C(0x1C000),
		UINT32_C(0x1C000), UINT32_C(0x1C000), UINT32_C(0x1C000), UINT32_C(0x1C000), UINT32_C(0x1C000), UINT32_C(0x1C000), UINT32_C(0x1C000), WITH_MB(0x38000),
		WITH_MB(0x38000), UINT32_C(0x54000), UINT32_C(0x54000), UINT32_C(0x54000), UINT32_C(0x54000), UINT32_C(0x54000), UINT32_C(0x54000), UINT32_C(0x54000),
		UINT32_C(0x54000), UINT32_C(0x54000), UINT32_C(0x54000), UINT32_C(0x54000), UINT32_C(0x54000), UINT32_C(0x54000), UINT32_C(0x54000), UINT32_C(0x54000),
		UINT32_C(0x54000), UINT32_C(0x54000), UINT32_C(0x54000), UINT32_C(0x54000), UINT32_C(0x54000), UINT32_C(0x54000), UINT32_C(0x54000), UINT32_C(0x54000),
		UINT32_C(0x54000), UINT32_C(0x54000), UINT32_C(0x54000), UINT32_C(0x54000), UINT32_C(0x54000), UINT32_C(0x54000), UINT32_C(0x54000), WITH_MB(0x70000)
	};

	uint32_t res = tbl[inpx >> 10];
```

**Comment by solidpixel** [comment]
Thanks for the bug report. Just so I can add a test case for it, how did you build the compressor, and what command line are you using?

**Comment by solidpixel** [comment]
FWIW, the fix you propose is incorrect - the "if" on line 167 will always succeed so everything gets treated as a "normal" value. The function will therefore fail to handle infinities, NaNs, and denormals.

**Comment by fb-aki** [comment]
Ah, you are correct...

**Comment by fb-aki** [comment]
line 167 should be changed to :
```
	if ((res & UINT32_C(0x80000000)) == 0)
```

**Comment by fb-aki** [comment]
Also, regarding how we are building astc-encoder, the compilation is actually hidden behind some tools and may take me a while to track down.  I do know this is a linux x86-64 machine and I'm using the ISA_SIMD=avx2 compiler flags and preprocess flags.

**Comment by solidpixel** [comment]
The `avx2` builds shouldn't be hitting this part of the code at all (all CPUs with AVX2 also support the F16C extensions, so they should be using that rather than the soft-float library).

**Comment by fb-aki** [comment]
Ah, good point.  I may have some compiler/preprocessor settings incorrect someplace.

**Comment by fb-aki** [comment]
This is a heavily sanitize version of what we are compiling with:

llvm/9.0.0/bin/clang++ -nostdinc -x c++ '-std=gnu++17' -faligned-new '-Wno-error=terminate' '-Wno-error=noexcept-type' '-Wno-error=format-truncation=' '-Wno-error=shadow=compatible-local' -Wno-class-memaccess -Wno-stringop-truncation -Wno-stringop-overflow '-Wno-error=redundant-move' '-Wno-error=address-of-packed-member' -Wnon-virtual-dtor -fopenmp '-finput-charset=UTF-8' -fsigned-char -Werror -Wall -Wno-comment -Wno-deprecated -Wno-deprecated-declarations -Wno-sign-compare -Wno-unused -Wunused-label -Wunused-result -Wno-builtin-macro-redefined -ffunction-sections -fdata-sections '-gz=none' '-march=corei7' '-mtune=skylake' -mpclmul '--gcc-toolchain=not_gcc_toolchain' -Qunused-arguments -Wgcc-compat -Wno-c++14-extensions -Wno-infinite-recursion -Wno-mismatched-tags -Wno-missing-braces -Wno-nullability-completeness '-Wno-error=undefined-var-template' '-Wno-error=conditional-uninitialized' '-Wno-error=constant-conversion' '-Wno-error=null-conversion' '-Wno-error=#warnings' -Wimport-preprocessor-directive-pedantic -Wthread-safety -Wno-address-of-packed-member -Wno-c++1z-compat -Wno-exceptions -Wno-unknown-warning-option -Wno-enum-compare -Wno-unused-private-field -Wno-unused-variable -Wno-return-std-move-in-c++11 -Wno-extra-semi-stmt -Wno-defaulted-function-deleted -Wno-implicit-const-int-float-conversion -mllvm '-profile-guided-section-prefix=false' '-Wno-error=backend-plugin' -Wno-address -Wno-unreachable-code-aggressive -Wno-inconsistent-missing-destructor-override -Wno-inconsistent-missing-override -Wno-self-assign -Wno-tautological-compare -Wno-constant-logical-operand -Wno-parentheses-equality -Wno-uninitialized-const-reference -gno-column-info -fcoroutines-ts -fdebug-types-section -fsized-deallocation '-fopenmp-version=31' '-ftemplate-depth=512' -g2 -gdwarf-aranges -gno-record-gcc-switches -DNDEBUG -DFBCODE_OPT_BUILD '-std=gnu++17' -faligned-new '-Wno-error=terminate' '-Wno-error=noexcept-type' '-Wno-error=format-truncation=' '-Wno-error=shadow=compatible-local' -Wno-class-memaccess -Wno-stringop-truncation -Wno-stringop-overflow '-Wno-error=redundant-move' '-Wno-error=address-of-packed-member' -Wnon-virtual-dtor -fopenmp '-finput-charset=UTF-8' -fsigned-char -Werror -Wall -Wno-comment -Wno-deprecated -Wno-deprecated-declarations -Wno-sign-compare -Wno-unused -Wunused-label -Wunused-result  -Wno-builtin-macro-redefined -ffunction-sections -fdata-sections '-gz=none' '-march=corei7' '-mtune=skylake' -mpclmul '--gcc-toolchain=not_gcc_toolchain' -Qunused-arguments -Wgcc-compat -Wno-c++14-extensions -Wno-infinite-recursion -Wno-mismatched-tags -Wno-missing-braces -Wno-nullability-completeness '-Wno-error=undefined-var-template' '-Wno-error=conditional-uninitialized' '-Wno-error=constant-conversion' '-Wno-error=null-conversion' '-Wno-error=#warnings' -Wimport-preprocessor-directive-pedantic -Wthread-safety -Wno-address-of-packed-member -Wno-c++1z-compat -Wno-exceptions -Wno-unknown-warning-option -Wno-enum-compare -Wno-unused-private-field -Wno-unused-variable -Wno-return-std-move-in-c++11 -Wno-extra-semi-stmt -Wno-defaulted-function-deleted -Wno-implicit-const-int-float-conversion -mllvm '-profile-guided-section-prefix=false' '-Wno-error=backend-plugin' -Wno-address -Wno-unreachable-code-aggressive -Wno-inconsistent-missing-destructor-override -Wno-inconsistent-missing-override -Wno-self-assign -Wno-tautological-compare -Wno-constant-logical-operand -Wno-parentheses-equality -Wno-uninitialized-const-reference -gno-column-info -fcoroutines-ts -fdebug-types-section -fsized-deallocation '-fopenmp-version=31' '-ftemplate-depth=512' -g2 -gdwarf-aranges -gno-record-gcc-switches -O3 -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -fdebug-info-for-profiling -fexperimental-new-pass-manager '-march=corei7' '-mtune=skylake' -mpclmul '-DASTCENC_NEON=0' '-DASTCENC_SSE=41' '-DASTCENC_AVX=2' '-DASTCENC_POPCNT=1' '-DASTCENC_F16C=1' '-DFOLLY_XLOG_STRIP_PREFIXES=""'  -w -Wno-deprecated-declarations -Wno-implicit-function-declaration -Wno-error -Wno-non-pod-varargs -Wno-format-security -Wno-c++11-narrowing -Wno-ignored-attributes -Wno-return-std-move -Wno-shadow -Wno-sign-compare -Wno-switch -Wno-undef -Wno-uninitialized -Wno-unknown-pragmas -Wno-unknown-warning-option -Wno-unused-function -Wno-unused-local-typedef -Wno-unused-value -Wno-unused-variable -Wno-register -Wno-format -Wno-unused-lambda-capture -Wno-missing-braces -Wno-unused-parameter -mf16c -mfma -mpopcnt -mavx -mavx2  -fdebug-info-for-profiling -Xclang -fdebug-compilation-dir -Xclang . -c -MD -MF astc-encoder/Source/astcenc_diagnostic_trace.cpp

Of course, this is after I fixed the compiler/preprocessor settings for proper AVX2 detection.

**Comment by solidpixel** [comment]
Any idea what option triggers the generation of a runtime error? Running on a CoffeeLake machine here, compiling with `clang++-10` on Linux (WSL) and I don't get a runtime error for the shift.

**Comment by solidpixel** [comment]
Found it - it can be triggered with `-fsanitize=undefined`

**Comment by solidpixel** [comment]
Should be fixed on `main` in commit 36469f9. Let me know how it goes.
